### ISSUE TITLE
holochain/bin

### ISSUE THREAD

**Issue by christopherreay** [metacurrency/holochain#204]
basic end user docs for this issue: https://github.com/metacurrency/holochain/wiki/Install-Holochain-on-nix

At the moment, the holochain bin scripts are inside the metacurrency/holochain github repo.
Once that repo is cloned onto the host machine, the scripts should allow the user to easily "install" holochain on the host system, and then manage their Holochain Core (on docker) installation, and manage their Holochain App development process

- [x] install holochain on the host system
- [x] initialise a holochain repository
- [x] run unit testing on a holochain repository
- [x] run integration, scenario testing on a holochain repository
- [x] add stuff to .gitignore during holochain.app.init
- [ ] port tests from holoSkel / make new ones
- [ ] really these scripts should be downloaded in a tarball or something, or otherwise "standalone", and should download the repo etc. when/if  needed (#174)
- [ ] playing around
- [ ] distribution

This will be implemented as bash scripts, and with reference also to https://github.com/metacurrency/holochain/wiki/Holochain-Development-Introduction

part of #174 

basically be like:
```bash
1. git clone metacurrency/holochain hc
2. hc/bin/holochain.system.install
3. source ~/.bin/.bashrc
4. cd /myHolochainApp
5. holochain.app.init
6. holochain.app.testScenario [tab tab]
7. nice
```

**Comment by christopherreay** [comment]
Bibliography:

* install the holochain bin
  * Where is the holochain bin?
    * https://stackoverflow.com/questions/59895/getting-the-source-directory-of-a-bash-script-from-within

**Comment by christopherreay** [comment]
holochain.init

921f5bdcec39046637aa160c3957d0246ff8e64e

Bibliography
* check if directory empty
  * https://www.cyberciti.biz/faq/linux-unix-shell-check-if-directory-empty/
* the "pass" command
  * https://stackoverflow.com/questions/17583578/what-command-means-do-nothing-in-a-conditional-in-bash

**Comment by christopherreay** [comment]
holochain.testScenario

start moving scripts over from the "top down"
e7f960708a45f50a9bb2bbdc055c30a11652be9e
d5df0ceff15c0a699e0e7ff5629e5389642b95d4

add runCluster and configuration for logging etc
349af60a7edafa5d9eea9298e1665249e2be8e49

Bibliography

**Comment by timotree3** [comment]
I'm a bit unclear on what this issue is implementing. Are you working on creating an easy packaging system for installing holochain apps?

If so I have a few thoughts, but I'll save them for if you say yes.

**Comment by christopherreay** [comment]
Ive updated the description. Is it clearer now?

**Comment by timotree3** [comment]
Not really. 😐

**Comment by christopherreay** [comment]
@timotree3  Im sorry. ive added a bit mot

**Comment by christopherreay** [comment]
holochain.system.bashCompletion

* called by .bashrc to add holochain bash completion to shell

3697d8e1987e45388a052ffec8d1dcd35a131e54
810e40305d2997a577e0cf60ca80c2f6569abbd4

**Comment by christopherreay** [comment]
TODO: add to .gitignore on `holochain.app.init`

**Comment by christopherreay** [comment]
@zippy merged and basic documentation here https://github.com/metacurrency/holochain/wiki/Install-Holochain-on-nix

**Comment by zippy** [comment]
I merged the stuff in install-Holochain-on-nix into https://github.com/metacurrency/holochain/wiki/Holochain-App-Development-Introduction

**Comment by zippy** [comment]
Convert these bash scripts to their own go program

**Comment by zippy** [comment]
So I woke up this morning thinking that to go program should be `hcdev` and we should move the `seed` and `test`commands from the regular `hc` into it too..

**Comment by zippy** [comment]
work on this (hcdev) happening in: https://github.com/metacurrency/holochain/tree/204-hcdev-cmd

**Comment by christopherreay** [comment]
RIght, so I am starting to implment calls to scripts in hcdev

First thing is a script for Core development that allows people to easliy compile and build the local source onto their system

**Comment by christopherreay** [comment]
@zippy so Ive gutted the entire of hcdev and hcdev_test... 
in the current master, running `go test" in cmd/hcdev/ fails with the output I sent you. Is this "supposed to happen" because of some pathing or something?

Irrespective of that, the Sprintf from this test here isnt functioning:
```bash
christopher@holochain:~/dev.bin2hcdcev/cmd/hcdev$ go test
x
Failures:

  * /home/christopher/dev.bin2hcdcev/cmd/hcdev/hcdev_test.go 
  Line 22:
  Expected: 'something'
  Actual:   '/tmp/go-build892150939/_/home/christopher/dev.bin2hcdcev/cmd/hcdev/_test'
  (Should be equal)


1 total assertion

--- FAIL: TestLocationOfExecutable (0.00s)
FAIL
exit status 1
FAIL    _/home/christopher/dev.bin2hcdcev/cmd/hcdev     0.007s
```

**Comment by zippy** [comment]
Ok so to be able to use `go test` you have to first do a gx-go rewrite from the base level.  Then go test from inside `cmd/hc` and `cmd/hcdev` should work.

**Comment by christopherreay** [comment]
@zippy thanks. Oh yeah. Thats not something that happens in dockerContainers. Cool that make life easier. Aye, because the import names are changed to ipfs hashes init

**Comment by christopherreay** [comment]
So, hcdev has three commands so far implemented in this branch:

1. hcdev core branch install
  * this checks that the current directory is at the root of metacurrency/holochain
  * and copies the files over $GOPATH/src/github.com/metacurrency/holochain
  * and then runs make bs, hc, hcdev
> resulting in an installation of the current checked out branch you are sitting in

> there should probably be a "reset" command to go back to master
And loads of other stuff, but hey. .Its versioning jim

2. hcdev app ini
  * interactively initialises a diretory to be a Holochain App directory
> end up with .hc directory with an App name in it

3. hcdev app testScenario --scenario=<scenarioName>
  * runs the testScenario script we all know and love
> ends up with loads of colorful output on the screen
and a file in logs/holochain/

**Comment by zippy** [comment]
looking pretty good.  A few things:

- I'd prefer to have the scenario specified as an arg rather than a flag because why type scenario twice.
- I'm not sure I like the whole "app" prefix, because that would imply that the current "test" and "web" commands (which you temporarily removed) would go as sub-commands under app.  Because the app commands are the most common, I think we could just assume default app context, and move "init" and "testScenario" to the top level, and then how about we rename "testScenario" to just "scenario" it kind of make sense no?

**Comment by christopherreay** [comment]
* In general I dont type anything, I use tab completion, and I prefer specific over implied any day
* for me the reality of using tools is that people move to fast and think to fast. I prefer to encourage them to think properly and not care at all about how long something takes to type, because if it is actually the right thing, then the speed they are going at is *very fast*
* `hcdev scenario` vs `hcdev app scenario` vs `hcdev app testScenario` I just dont like writing things that people can only understand onces they already are part of your ecosystem.
* I agree about the flag.

**Comment by zippy** [comment]
ok so I just tried this: `hcdev core branch install` and I got:

```
HC: core.branch.install: installing from: /home/eric/go/src/github.com/metacurrency/holochain
this will remove the existing content at /home/eric/go/src/github.com/metacurrency/holochain (y/N) y
/home/eric/go/src/github.com/metacurrency/holochain/bin/holochain.core.testing.branch: line 13: /home/eric/go/src/github.com/metacurrency/holochain/bin/holochain.tools.confirm: No such file or directory
HC: exiting
```
This left things in a bad state because my directory got emptied, but not completely deleted which meant that a `go get -d github.com/metacurrency/holochain` failed until I manually went and removed the directory!

**Comment by christopherreay** [comment]
1. yeah. It doesnt make any sense to run that command from there. Ill put in a check for it.
2. Also the fact that bin/holochain.system.install isnt installed is on the todo list for stuff that is currently "broken".
  * if bin/holochain.... is installed, it works, with the caveat above

**Comment by christopherreay** [comment]
Actually the script holochain.core.testing.branch isnt very well written. I was a bit swamped in addressing the go side of it to bother properly with the script. 

Ill put as many checks as possible in the go code before runing the scripts. But most of the scripts are very defensively written. I think all of them apart from that one would fail immediately on running

**Comment by zippy** [comment]
why should you have to run `hcdev core branch install` from anywhere in particular.  It seems like the context is global singular, so it should just happen to GOPATH  no matter where you run that command from.  The app parts of the command are contextual so it makes sense that they would assume the current directory as the context (or, perhaps we also add a --path option so you can specify it explicity)

**Comment by christopherreay** [comment]
> why should you have to run hcdev core branch install from anywhere in particular

because it doesnt checkout a branch or anything like that, it just copies the local source files over the GOPATH version of metacurrency/holochain, and runs make etc

>. It seems like the context is global singular, so it should just happen to GOPATH no matter where you run that command from. 

I dont think that is from the same point of view as that thing is for from my point of view :)

> The app parts of the command are contextual so it makes sense that they would assume the current directory as the context (or, perhaps we also add a --path option so you can specify it explicity)

thats exactly how it works now

**Comment by christopherreay** [comment]
There's no reason why it couldnt just use git to download a branch over the top of $GOPATH and then install. I can do that. Obvs thats what I was thinking, I was just working on the interface between go and the scripts.

**Comment by zippy** [comment]
ok, got it...  yah currently that script does this check:

`[ ! -f ./bin/holochain.system.checkInstalled ] && { echo "HC: not in a holochain core directory. exitting" && exit 1 ; }`

which I assume is to see if the current directory is a holochain app, and then tries to copy that app into the GOPATH?  :-P

**Comment by zippy** [comment]
regarding: "I just dont like writing things that people can only understand onces they already are part of your ecosystem."  I think maybe I just need to be retrained or something, because for me meaning comes from compression.  And when things get too long I loose the meaning.  I find it much more confusing and hard to work with.  I get that verbose stuff is easier to learn about, but I'm interested in what's easiest to work with.  And I think I'm not usual in that regard.  Fortunately "github.com/urfave/cli" allows for aliases for commands, so maybe we can come up with something that will satisfy both of us...

**Comment by christopherreay** [comment]
Id also rename hc to holochain
and have hcdev as a subcommand so
`holochain development app testScenario SCENARIONAME`
or
`holochain developer app testScenario SCENARIONAME`

**Comment by zippy** [comment]
Arg! That would drive me totally nuts!

**Comment by christopherreay** [comment]
what about
`hc dapp scenario SCENARIONAME` Thats cool

**Comment by zippy** [comment]
I don't see the difference between that and:

`hcdev scenario SCENARIONAME`

they seem totally equivalent.

**Comment by christopherreay** [comment]
one of them has the word "app" in it, which ties in with all the documentation
the otherone clearly shows psychotic, sociopathic and possibly demented tendencies

**Comment by zippy** [comment]
ah, now I understand.  That makes it all so clear!

**Comment by christopherreay** [comment]
so `hc` == `holochain` is the primary namespace
`dapp` and `dcore` both compile nicely to both
* development app/core
and
* distributed app/core
which is nice
and scenario in the context of holochain dapp is sufficiently expressive

if you must know :)
